feat: added IAsyncEnumerable<T> query overloads to QueryAPI

ritchiecarroll opened the pull request:
This adds [`IAsyncEnumerable<T>`](https://docs.microsoft.com/en-us/archive/msdn-magazine/2019/november/csharp-iterating-with-async-enumerables-in-csharp-8) support to the QueryAPI for .NET Standard 2.1 deployments. To support .NET Standard 2.1, client and client projects were modified to dual target .NET standard 2.1 and 2.0.

Closes #119 - both timestamp issue and suggested `IAsyncEnumerable` feature

bednar commented on Client.Core/Client.Core.csproj:
```diff
@@ -1,7 +1,7 @@
 
     <PropertyGroup>
-        <TargetFramework>netstandard2.0</TargetFramework>
+        <TargetFrameworks>netstandard2.1;netstandard2.0</TargetFrameworks>
```
1. Please change order of versions and add `LangVersion`:
```suggestion
        <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
        <LangVersion>8</LangVersion>
```
2. Add `Microsoft.Bcl.AsyncInterfaces` ass conditional reference:
```xml
<ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0' ">
        <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="1.1.1" />
</ItemGroup>
```

bednar commented on Client/Client.csproj:
```diff
@@ -1,9 +1,9 @@
 
     <PropertyGroup>
-        <TargetFramework>netstandard2.0</TargetFramework>
+      <TargetFrameworks>netstandard2.1;netstandard2.0</TargetFrameworks>
```
Please change order of versions and add LangVersion:

```suggestion
      <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
      <LangVersion>8</LangVersion>  
```

bednar reviewed (changes requested):
Hi @ritchiecarroll,

thanks for your PR!

Please add tests for your changes to be sure that everything works correctly.

Regards

ritchiecarroll commented:
Requested changes and tests have been added to PR

ritchiecarroll commented:
OK - I think I've done what I can here. Let me know how this looks.

bednar commented on Client.Core/Flux/Internal/FluxCsvParser.cs:
```diff
@@ -72,119 +76,156 @@ public void ParseFluxResponse(Stream source, ICancellable cancellable, IFluxResp
+            }
+        }
+
+#if NETSTANDARD2_1
```
Please remove this directive.

bednar commented on Client.Core/Internal/AbstractQueryClient.cs:
```diff
@@ -103,6 +105,27 @@ void Consumer(ICancellable cancellable, Stream bufferedStream)
             }
         }
 
+#if NETSTANDARD2_1
```
Please remove this directive.

bednar commented on Client.Legacy.Test/FluxCsvParserTest.cs:
```diff
@@ -428,6 +431,30 @@ public void ParsingToConsumer()
             Assert.That(records.Count == 2);
         }
 
+#if NETCOREAPP3_1
```
Please remove this directive.

bednar commented on Client.Legacy.Test/FluxCsvParserTest.cs:
```diff
@@ -456,6 +483,35 @@ public void CancelParsing()
             Assert.That(records.Count == 1);
         }
 
+#if NETCOREAPP3_1
```
Please remove this directive.

bednar commented on Client.Core/Client.Core.csproj:
```diff
@@ -38,5 +39,4 @@
       <PackageReference Include="RestSharp" Version="106.6.10" />
       <PackageReference Include="System.Net.Http" Version="4.3.4" />
     </ItemGroup>
-    
```
Please add conditional dependency to `Microsoft.Bcl.AsyncInterfaces`:

```xml
<ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0' ">
        <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="1.1.1" />
</ItemGroup>
```

ritchiecarroll replied:
> 
> Could you give me a permission to update your `GridProtectionAlliance:feat/AsyncEnumerable` branch? I will fix our CircleCI scripts to supports dual targeting.
> 
> Regards

Write permissions granted. Will attempt your other requests.

bednar reviewed (changes requested):
Hi @ritchiecarroll,

there is a few things to change:

1. `Client.Legacy.csproj` should also be a dual targeting:
```
<TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
<LangVersion>8</LangVersion>
```

2. We could remove `#if NETSTANDARD2_1`and `#if NETCOREAPP3_1` by adding optional dependency to `Client.Core.csproj`:
```
<ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0' ">
     <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="1.1.1" />
</ItemGroup>
```

Could you give me a permission to update your `GridProtectionAlliance:feat/AsyncEnumerable` branch? I will fix our CircleCI scripts to supports dual targeting.

Regards

ritchiecarroll commented:
> Hi @ritchiecarroll,
> 
> there is a few things to change:
> 
> 1. `Client.Legacy.csproj` should also be a dual targeting:
> 
> ```
> <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
> <LangVersion>8</LangVersion>
> ```
> 
> 1. We could remove `#if NETSTANDARD2_1`and `#if NETCOREAPP3_1` by adding optional dependency to `Client.Core.csproj`:
> 
> ```
> <ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0' ">
>      <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="1.1.1" />
> </ItemGroup>
> ```
> 
> Could you give me a permission to update your `GridProtectionAlliance:feat/AsyncEnumerable` branch? I will fix our CircleCI scripts to supports dual targeting.
> 
> Regards

All requests completed. Also, check your e-mail associated with your GitHub account, you should have a pending invite for write permissions to the `GridProtectionAlliance:feat/AsyncEnumerable`  branch so that you can adjust CircleCI scripts as needed. I appreciate that step as I otherwise have to provide permissions to this tool for "all" repos I have write access to, which is more risk than I would like to incur until I trust this CI tool more.

Thanks,
Ritchie

ritchiecarroll commented:
Hi @bednar - anything remaining for me to do here?

Thanks,
Ritchie

bednar reviewed (changes requested):
Hi @ritchiecarroll, I've fixed CircleCI scripts to build your branch. 

Could you please rebase your PR with master and fix these failing tests: `InfluxDB.Client.Test.ItWriteQueryApiTest.WriteMeasurements`, `InfluxDB.Client.Test.ItWriteQueryApiTest.WriteMeasurementsWithoutFields`?
    
- https://app.circleci.com/pipelines/github/influxdata/influxdb-client-csharp/433/workflows/c658cc41-849b-46ca-8126-015b177fd560/jobs/1379

and again Thanks for you contribution.

ritchiecarroll commented:
> Hi @ritchiecarroll, I've fixed CircleCI scripts to build your branch.
> 
> Could you please rebase your PR with master and fix these failing tests: `InfluxDB.Client.Test.ItWriteQueryApiTest.WriteMeasurements`, `InfluxDB.Client.Test.ItWriteQueryApiTest.WriteMeasurementsWithoutFields`?
> 
> * https://app.circleci.com/pipelines/github/influxdata/influxdb-client-csharp/433/workflows/c658cc41-849b-46ca-8126-015b177fd560/jobs/1379
> 
> and again Thanks for you contribution.

OK @bednar , fixed those.

FYI, I think the remaining failure on `dotnet-3.1-nightly` seems more related to this exception:
```
InfluxDB.Client.Core.Exceptions.HttpException : retention policy duration must be at least 1h0m0s
```
Is this a test configuration issue?

Thanks.
Ritchie

bednar commented:
The remaining failure on `dotnet-3.1-nightly` seems to be related to nightly build of InfluxDB.

Please don't use a marge commits from master, because the PR contains unrelated changes. 

You should use rebase feature. This commands fix your git history:

```console
git remote add upstream https://github.com/influxdata/influxdb-client-csharp/
git fetch upstream

# reset before merge commit
git reset --hard 47e9727640976dcbaf0cb6f14af3c90bdac1d0eb
# rebase from master
git rebase upstream/master
# force push
git push -f origin feat/AsyncEnumerable
```

Thanks

ritchiecarroll commented:
> The remaining failure on `dotnet-3.1-nightly` seems to be related to nightly build of InfluxDB.
> 
> Please don't use a marge commits from master, because the PR contains unrelated changes.
> 
OK @bednar, successful build after applying your specific rebase steps.

Thanks,
Ritchie

ritchiecarroll commented:
@bednar: FYI, after successful build, the PR says "Merging can be performed automatically once the requested changes are addressed." but I don't think there is something I can do here, I believe this may be a step for you:

Changes requested
1 review requesting changes by reviewers with write access.

bednar commented on Client.Core/Flux/Internal/FluxResultMapper.cs:
```diff
@@ -106,85 +110,67 @@ private void SetFieldValue<T>(T poco, PropertyInfo property, object value)
+                    $"The correct type is '{value.GetType().Name}' (current field value: '{value}'). Exception: {ex.Message}", ex);
             }
         }
 
-        private double ToDoubleValue(object value)
+        private DateTime ToDateTimeValue(object value)
         {
-            if (value.GetType().IsAssignableFrom(typeof(double)))
+            if (value is DateTime dateTime)
             {
-                return (double) value;
+                return dateTime;
             }
 
-            return (double) value;
-        }
```
The following statement is unnecessary in `ToDateTimeValue`:

```csharp
if (value is DateTime dateTime)
{
      return dateTime;
}
```

ritchiecarroll replied:
I do not feel this is true.  poco property may be `DateTime`, but object value coming may not be `DateTime`

ritchiecarroll replied:
Or are you saying because of last line that already does something similar?

ritchiecarroll replied:
My thinking was this:

1) Handle case where value is of type `DateTime`
2) Handle case where value is of type `Instant`
3) Last chance, attempt cast to `DateTime` in case there are implict or explicit conversions available to convert to `DateTime` with associated value instance

ritchiecarroll replied:
OK, I updated implementation to fail appropriately when types do not match.

bednar replied:
Nice! Perfect improvement 👍

bednar commented on Client.Core/Flux/Internal/FluxResultMapper.cs:
```diff
@@ -106,85 +110,67 @@ private void SetFieldValue<T>(T poco, PropertyInfo property, object value)
         }
 
-        private int ToIntValue(object value)
+        private Instant ToInstantValue(object value)
```
The following statement is unnecessary in `ToInstantValue`:

```csharp
if (value is Instant instant)
{
     return instant;
}
```

ritchiecarroll replied:
I do not feel this is true.  poco property may be `Instant`, but object value coming may not be `Instant`

ritchiecarroll replied:
(see comment above for justification)

bednar reviewed (changes requested):
@ritchiecarroll we should also review an implementation, but we are on track to merge this to next release.

Thanks for your fantastic job here

bednar commented on Client.Core/Flux/Internal/FluxCsvParser.cs:
```diff
@@ -1,7 +1,11 @@
 using System.IO;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
```
Using `System.Threading.Tasks` is not required by the code and can be safely removed

```suggestion
```

bednar commented on Client.Core/Flux/Internal/FluxCsvParser.cs:
```diff
@@ -72,119 +76,154 @@ public void ParseFluxResponse(Stream source, ICancellable cancellable, IFluxResp
 
-            var parsingState = ParsingState.Normal;
+            using var csv = new CsvReader(new StreamReader(source), CultureInfo.InvariantCulture);
+            ParseFluxResponseState state = new ParseFluxResponseState { csv = csv };
```
Use 'var' (simple types):

```suggestion
            var state = new ParseFluxResponseState { csv = csv };
```

bednar commented on Client.Core/Flux/Internal/FluxCsvParser.cs:
```diff
@@ -72,119 +76,154 @@ public void ParseFluxResponse(Stream source, ICancellable cancellable, IFluxResp
+            Arguments.CheckNotNull(reader, nameof(reader));
+
+            using var csv = new CsvReader(reader, CultureInfo.InvariantCulture);
+            ParseFluxResponseState state = new ParseFluxResponseState { csv = csv };
```
Use 'var' (simple types):

```suggestion
            var state = new ParseFluxResponseState { csv = csv };
```

ritchiecarroll commented on Client.Core/Flux/Internal/FluxResultMapper.cs:
```diff
@@ -106,85 +110,72 @@ private void SetFieldValue<T>(T poco, PropertyInfo property, object value)
-                    return;
+                    // Nullable types cannot be used in type conversion, but we can use Nullable.GetUnderlyingType()
+                    // to determine whether the type is nullable and convert to the underlying type instead
+                    Type targetType = Nullable.GetUnderlyingType(propertyType) ?? propertyType;
```
I did a full update in code for converting specific type usage to `var` using ReSharper - there are many hundreds of cases that affects 251 files. I will update these per your request, if you would like me to replace all specific type instances with `var` in the entire project, may want to do that in a separate PR.

ritchiecarroll replied:
Also, you can make a specific code styling normative by adding an [`.editorconfig`](https://docs.microsoft.com/en-us/visualstudio/ide/create-portable-custom-editor-options?view=vs-2019) file to your project.

See here for enforcing use of `var`:
https://docs.microsoft.com/en-us/visualstudio/ide/editorconfig-language-conventions?view=vs-2019#csharp_style_var_for_built_in_types

To help contributors understand the rules, here is an example [coding-style markdown](https://github.com/gemstone/common/blob/master/docs/coding-style.md) that can help verbally codify the pattern too.

bednar replied:
Yeah, there are a lot of code style problems with autogenerated sources - `InfluxDB.Client.Api`... 😞 
Let's finish this PR and later we will see...

bednar reviewed (changes requested):
@ritchiecarroll,

Could you please update `CHANGELOG.md`? After that we will be ready to merge it into master.

Thanks for your fantastic job! 💯

ritchiecarroll commented:
Change log updated.

ritchiecarroll commented:
@bednar just want to make sure you are not waiting on me to do something here. Thanks!

bednar commented:
Hi @ritchiecarroll, the PR is merged into master. 

If you would like to use a preview version use our nightly build:

```
dotnet add package InfluxDB.Client --version 1.13.0-dev.1489
```

Regards
